File: server/lib/state.js
/* Builds the role-scoped state object the browser app renders from. Customers only ever receive their own data. */
const { db, getSetting } = require('../db');
const S = require('./serialize');
const { j } = require('./util');
const pay = require('../services/payments');
const { expireUnpaid } = require('../services/bookings');

const mask = (m) => (m ? m.slice(0, 2) + 'XXXXXX' + m.slice(-2) : '');

function buildState(auth) {
  expireUnpaid();
  const role = auth && auth.role;
  const kits = db.prepare('SELECT * FROM kits').all();
  const pujaRows = db.prepare(role === 'admin' ? 'SELECT * FROM pujas' : 'SELECT * FROM pujas WHERE hidden=0').all();
  const kitItems = Object.fromEntries(kits.map((k) => [k.id, j(k.items, [])]));

  const st = {
    session: auth ? { role, uid: auth.uid, pid: auth.pid || null } : null,
    config: { paymentMode: pay.mode(), razorpayKeyId: pay.mode() === 'razorpay' ? process.env.RAZORPAY_KEY_ID : null, demo: String(process.env.DEMO_MODE || (process.env.NODE_ENV === 'production' ? 'false' : 'true')) === 'true' },
    catalog: {
      pujas: pujaRows.map((r) => S.puja(r, kitItems[r.kit])), kits: kits.map(S.kit),
      prasad: db.prepare('SELECT * FROM prasad').all().map(S.prasad), temples: db.prepare('SELECT * FROM temples').all().map(S.temple),
      festivals: db.prepare('SELECT * FROM festivals ORDER BY date').all().map(S.festival)
    },
    banners: db.prepare('SELECT * FROM banners WHERE enabled=1').all().map((b) => ({ id: b.id, t: b.text, on: true })),
    kundali: { enabled: true, purposes: ['General', 'Marriage', 'Career', 'Business', 'Health & Wellness', 'Finance', 'Education', 'Family', 'Child', 'Spiritual', 'Property', 'Other'] },
    toggles: (() => { try { const r = db.prepare("SELECT value FROM settings WHERE key='service_toggles'").get(); return Object.assign({ home: true, online: true, temple: true, customized: true, kundali: true, pandit: true, templeDir: true, prasad: true, samagri: true, astrology: true }, r ? JSON.parse(r.value) : {}); } catch (e) { return {}; } })(),
    pandits: [], busy: [], reviews: [],
    me: null, users: [], bookings: [], orders: [], notifs: [], tickets: [], coupons: [], payouts: [], inv: {}, campaigns: [], leads: [], set: {}, hidden: []
  };

  const pRows = db.prepare('SELECT * FROM pandits').all();
  st.pandits = pRows.filter((p) => role === 'admin' || p.status === 'verified' || (role === 'pandit' && p.id === auth.pid)).map((p) => S.pandit(p, { admin: role === 'admin', self: role === 'pandit' && p.id === auth.pid }));
  st.busy = db.prepare("SELECT id, pandit_id p, date, slot FROM bookings WHERE pandit_id IS NOT NULL AND status NOT IN ('Cancelled')").all();
  st.reviews = db.prepare("SELECT b.pandit_id pid, b.review, b.puja_id FROM bookings b WHERE b.review IS NOT NULL AND b.review_hidden=0 AND b.pandit_id IS NOT NULL").all()
    .map((r) => { const rv = j(r.review, {}); return { pid: r.pid, r: rv.r, t: rv.t, by: rv.by, puja: r.puja_id }; });

  if (role === 'customer') {
    const u = db.prepare('SELECT * FROM users WHERE id=?').get(auth.uid);
    st.me = S.user(u); st.users = [st.me];
    /* Family members live in the family_members table (single source of truth);
       `fam` keeps its legacy shape for the booking-wizard member picker. */
    st.me.fam = db.prepare('SELECT * FROM family_members WHERE customer_id=? ORDER BY created_at').all(u.id)
      .map((f) => ({ id: f.id, n: f.name, rel: f.relationship, gotra: f.gotra }));
    st.family = db.prepare('SELECT * FROM family_members WHERE customer_id=? ORDER BY created_at').all(u.id)
      .map((f) => ({ id: f.id, relationship: f.relationship, name: f.name, gender: f.gender, dob: f.dob, tob: f.tob, birthPlace: f.birth_place, city: f.city, state: f.state, country: f.country, lat: f.lat, lon: f.lon, tz: f.tz, gotra: f.gotra, notes: f.notes, createdAt: f.created_at }));
    st.myKundalis = db.prepare('SELECT id,name,relationship,billing,price,gst,final_amount,payment_status,order_id,created_at FROM kundalis WHERE customer_id=? ORDER BY created_at DESC').all(u.id)
      .map((k) => ({ kundaliId: k.id, name: k.name, relationship: k.relationship || 'Self', billing: k.billing, price: k.price, gst: k.gst, final: k.final_amount, paymentStatus: k.payment_status, orderId: k.order_id, createdAt: k.created_at }));
    st.bookings = db.prepare("SELECT * FROM bookings WHERE user_id=? AND status != 'PendingPayment' ORDER BY created DESC").all(u.id).map(S.booking);
    st.orders = db.prepare('SELECT * FROM orders WHERE user_id=? ORDER BY date DESC').all(u.id).map(S.order);
    st.notifs = db.prepare('SELECT * FROM notifs WHERE user_id=? ORDER BY ts DESC LIMIT 100').all(u.id).map((n) => ({ id: n.id, uid: n.user_id, ch: n.channel, m: n.message, ts: n.ts }));
    st.tickets = db.prepare('SELECT * FROM tickets WHERE user_id=? ORDER BY rowid DESC').all(u.id).map(S.ticket);
  } else if (role === 'pandit') {
    st.bookings = db.prepare("SELECT * FROM bookings WHERE pandit_id=? AND status != 'PendingPayment' ORDER BY date").all(auth.pid).map(S.booking);
    const ids = [...new Set(st.bookings.map((b) => b.userId))];
    st.users = ids.map((id) => db.prepare('SELECT id,name,mobile FROM users WHERE id=?').get(id)).filter(Boolean).map((u) => ({ id: u.id, n: u.name, m: mask(u.mobile), e: '', pts: 0, plus: false, addr: [], fam: [], pref: {}, joined: '' }));
    st.payouts = db.prepare('SELECT * FROM payouts WHERE pandit_id=?').all(auth.pid).map(S.payout);
    st.set = { comm: getSetting('commission', 20) };
  } else if (role === 'admin') {
    st.bookings = db.prepare("SELECT * FROM bookings WHERE status != 'PendingPayment' ORDER BY created DESC").all().map(S.booking);
    st.users = db.prepare("SELECT * FROM users WHERE role='customer'").all().map(S.user);
    st.orders = db.prepare('SELECT * FROM orders ORDER BY date DESC').all().map(S.order);
    st.tickets = db.prepare('SELECT * FROM tickets ORDER BY rowid DESC').all().map(S.ticket);
    st.coupons = db.prepare('SELECT * FROM coupons').all().map(S.coupon);
    st.payouts = db.prepare('SELECT * FROM payouts').all().map(S.payout);
    st.inv = Object.fromEntries(kits.map((k) => [k.id, k.stock]));
    st.campaigns = db.prepare('SELECT * FROM campaigns').all().map((c) => ({ id: c.id, n: c.name, ch: c.channel, aud: c.audience, st: c.status, sent: c.sent }));
    st.leads = db.prepare('SELECT * FROM leads ORDER BY id DESC LIMIT 200').all().map((l) => ({ type: l.type, n: l.name, d: l.details, date: l.date }));
    st.set = { comm: getSetting('commission', 20) };
    st.hidden = db.prepare('SELECT id FROM bookings WHERE review_hidden=1').all().map((r) => r.id);
    st.banners = db.prepare('SELECT * FROM banners').all().map((b) => ({ id: b.id, t: b.text, on: !!b.enabled }));
  }
  return st;
}
module.exports = { buildState };

File: backend-python/app/state.py
"""Role-scoped state builder — port of server/lib/state.js. Customers only ever
receive their own data; pandits see masked mobiles; admins see everything."""
import os

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from .services import payments as pay
from .models import (Banner, Booking, Coupon, Festival, Kit, Notif, Order,
                     Pandit, Prasad, Payout, Puja, Temple, Ticket, User)
from .serialize import (booking, coupon, festival, kit, notif, order, pandit,
                        prasad, puja, temple, ticket, payout as s_payout, user)
from .services.bookings import expire_unpaid, get_setting
from .util import j


def _mask(m: str | None) -> str:
    return (m[:2] + "XXXXXX" + m[-2:]) if m else ""


async def build_state(db: AsyncSession, auth: dict | None) -> dict:
    await expire_unpaid(db)
    role = auth and auth.get("role")
    kits = (await db.execute(select(Kit))).scalars().all()
    puja_rows = (await db.execute(
        select(Puja) if role == "admin" else select(Puja).where(Puja.hidden == 0))).scalars().all()
    kit_items = {k.id: j(k.items, []) for k in kits}
    demo = (os.environ.get("DEMO_MODE") or "true").lower() == "true"

    st = {
        "session": ({"role": role, "uid": auth.get("uid"), "pid": auth.get("pid") or None}
                    if auth else None),
        "config": {"paymentMode": pay.mode(),
                   "razorpayKeyId": os.environ.get("RAZORPAY_KEY_ID") if pay.mode() == "razorpay" else None,
                   "demo": demo},
        "catalog": {
            "pujas": [puja(r, kit_items.get(r.kit)) for r in puja_rows],
            "kits": [kit(k) for k in kits],
            "prasad": [prasad(k) for k in (await db.execute(select(Prasad))).scalars().all()],
            "temples": [temple(t) for t in (await db.execute(select(Temple))).scalars().all()],
            "festivals": [festival(f) for f in (await db.execute(
                select(Festival).order_by(Festival.date))).scalars().all()],
        },
        "banners": [{"id": b.id, "t": b.text, "on": True} for b in
                    (await db.execute(select(Banner).where(Banner.enabled == 1))).scalars().all()],
        "kundali": {"enabled": True, "purposes": ["General", "Marriage", "Career", "Business",
                                                  "Health & Wellness", "Finance", "Education",
                                                  "Family", "Child", "Spiritual", "Property", "Other"]},
        "toggles": {"home": True, "online": True, "temple": True, "customized": True,
                    "kundali": True, "pandit": True, "templeDir": True, "prasad": True,
                    "samagri": True, "astrology": True},
        "pandits": [], "busy": [], "reviews": [],
        "me": None, "users": [], "bookings": [], "orders": [], "notifs": [], "tickets": [],
        "coupons": [], "payouts": [], "inv": {}, "campaigns": [], "leads": [], "set": {}, "hidden": [],
    }
    _ = get_setting  # imported for parity; settings surfaced per-role below

    p_rows = (await db.execute(select(Pandit))).scalars().all()
    st["pandits"] = [pandit(p, admin=(role == "admin")) for p in p_rows
                     if role == "admin" or p.status == "verified"
                     or (role == "pandit" and p.id == auth.get("pid"))]
    busy_rows = (await db.execute(select(Booking.id, Booking.pandit_id, Booking.date, Booking.slot).where(
        Booking.pandit_id.isnot(None), Booking.status != "Cancelled"))).all()
    st["busy"] = [{"id": r.id, "p": r.pandit_id, "date": r.date, "slot": r.slot} for r in busy_rows]
    st["reviews"] = [{"pid": r.pandit_id, "r": j(r.review, {}).get("r"),
                      "t": j(r.review, {}).get("t"), "by": j(r.review, {}).get("by"),
                      "puja": r.puja_id} for r in (await db.execute(
        select(Booking).where(Booking.review.isnot(None), Booking.review_hidden == 0,
                              Booking.pandit_id.isnot(None)))).scalars().all()]

    if role == "customer":
        u = await db.get(User, auth["uid"])
        st["me"] = user(u)
        st["users"] = [st["me"]]
        st["bookings"] = [booking(b) for b in (await db.execute(
            select(Booking).where(Booking.user_id == u.id, Booking.status != "PendingPayment")
            .order_by(Booking.created.desc()))).scalars().all()]
        st["orders"] = [order(o) for o in (await db.execute(
            select(Order).where(Order.user_id == u.id).order_by(Order.date.desc()))).scalars().all()]
        st["notifs"] = [notif(n) for n in (await db.execute(
            select(Notif).where(Notif.user_id == u.id).order_by(Notif.ts.desc())
            .limit(100))).scalars().all()]
        st["tickets"] = [ticket(t) for t in (await db.execute(
            select(Ticket).where(Ticket.user_id == u.id).order_by(Ticket.id.desc()))).scalars().all()]
    elif role == "pandit":
        st["bookings"] = [booking(b) for b in (await db.execute(
            select(Booking).where(Booking.pandit_id == auth["pid"],
                                  Booking.status != "PendingPayment")
            .order_by(Booking.date))).scalars().all()]
        ids = list(dict.fromkeys(b["userId"] for b in st["bookings"]))
        st["users"] = [{"id": uid, "n": (u2.name if (u2 := await db.get(User, uid)) else ""),
                        "m": _mask(u2.mobile if u2 else ""), "e": "", "pts": 0, "plus": False,
                        "addr": [], "fam": [], "pref": {}, "joined": ""} for uid in ids]
        st["payouts"] = [s_payout(p) for p in (await db.execute(
            select(Payout).where(Payout.pandit_id == auth["pid"]))).scalars().all()]
        st["set"] = {"comm": await get_setting(db, "commission", 20)}
    elif role == "admin":
        st["bookings"] = [booking(b) for b in (await db.execute(
            select(Booking).where(Booking.status != "PendingPayment")
            .order_by(Booking.created.desc()))).scalars().all()]
        st["users"] = [user(r) for r in (await db.execute(
            select(User).where(User.role == "customer"))).scalars().all()]
        st["orders"] = [order(o) for o in (await db.execute(
            select(Order).order_by(Order.date.desc()))).scalars().all()]
        st["tickets"] = [ticket(t) for t in (await db.execute(
            select(Ticket).order_by(Ticket.id.desc()))).scalars().all()]
        st["coupons"] = [coupon(c) for c in (await db.execute(select(Coupon))).scalars().all()]
        st["payouts"] = [s_payout(p) for p in (await db.execute(select(Payout))).scalars().all()]
        st["inv"] = {k.id: k.stock for k in kits}
        st["campaigns"] = []
        st["leads"] = []
        st["set"] = {"comm": await get_setting(db, "commission", 20)}
        st["hidden"] = [r.id for r in (await db.execute(
            select(Booking).where(Booking.review_hidden == 1))).scalars().all()]
        st["banners"] = [{"id": b.id, "t": b.text, "on": bool(b.enabled)} for b in
                         (await db.execute(select(Banner))).scalars().all()]
    return st
